File: app/vortasks/lib/models/shop/product.dart
class Product {
  final int id;
  final String name;
  final String description;
  final String icon;
  final int coins;
  final int gems;
  final int totalSales;

  Product({
    required this.id,
    required this.name,
    required this.description,
    required this.icon,
    required this.coins,
    required this.gems,
    required this.totalSales,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      icon: json['icon'],
      coins: json['coins'],
      gems: json['gems'],
      totalSales: json['totalSales'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'icon': icon,
      'coins': coins,
      'gems': gems,
      'totalSales': totalSales,
    };
  }

  Product copyWith({
    int? id,
    String? name,
    String? description,
    String? icon,
    int? coins,
    int? gems,
    int? totalSales,
  }) {
    return Product(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      icon: icon ?? this.icon,
      coins: coins ?? this.coins,
      gems: gems ?? this.gems,
      totalSales: totalSales ?? this.totalSales,
    );
  }
}

File: api/src/main/java/com/leandroadal/vortasks/entities/shop/product/Product.java
package com.leandroadal.vortasks.entities.shop.product;

import java.util.HashSet;
import java.util.Set;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@Entity
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "product")
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @Column(unique = true)
    private String name;

    private String description;
    private String icon;
    private Integer coins;
    private Integer gems;
    private Integer totalSales;
    private Boolean active;

    @JoinTable(
      name = "product_category", 
      joinColumns = @JoinColumn(name = "product_id"), 
      inverseJoinColumns = @JoinColumn(name = "category_id"))
    @ManyToMany(cascade = CascadeType.ALL)
    @Setter(AccessLevel.NONE)
    private Set<Category> categories = new HashSet<>();

    public void sell(){
        this.totalSales++;
    }
}
